salt/utils/vault/factory.py

```diff
@@ -0,0 +1,1024 @@
+                    )
+                else:
+                    if unwrapped_response.get("auth"):
+                        result.update({"auth": unwrapped_response["auth"]})
+                    if unwrapped_response.get("data"):
+                        result.update({"data": unwrapped_response["data"]})
+
+        if config_expired:
+            raise VaultConfigExpired()
+
+        for key, val in misc_data.items():
+            tgt = "data" if result.get("data") is not None else "auth"
+            if (
+                salt.utils.data.traverse_dict_and_list(result, f"{tgt}:{key}", NOT_SET)
+                == NOT_SET
+            ):
+                salt.utils.dictupdate.set_dict_key_value(
+                    result,
+                    f"{tgt}:{key}",
+                    val,
+                )
+
+        result.pop("wrap_info", None)
+        result.pop("wrap_info_nested", None)
+        result.pop("misc_data", None)
+        return result, unwrap_client
+
+    global __salt__  # pylint: disable=global-statement
+    if not __salt__:
+        __salt__ = salt.loader.minion_mods(opts)
+
+    minion_id = opts["grains"]["id"]
+    pki_dir = opts["pki_dir"]
+
+    # When rendering pillars, the module executes on the master, but the token
+    # should be issued for the minion, so that the correct policies are applied
+    if opts.get("__role", "minion") == "minion":
+        private_key = f"{pki_dir}/minion.pem"
+        log.debug(
+            "Running on minion, signing request `vault.%s` with key %s",
+            func,
+            private_key,
+        )
+        signature = base64.b64encode(salt.crypt.sign_message(private_key, minion_id))
+        arg = [
+            ("minion_id", minion_id),
+            ("signature", signature),
+            ("impersonated_by_master", False),
+        ] + list(kwargs.items())
+
+        result = __salt__["publish.runner"](
+            f"vault.{func}", arg=[{"__kwarg__": True, k: v} for k, v in arg]
+        )
+    else:
+        private_key = f"{pki_dir}/master.pem"
+        log.debug(
+            "Running on master, signing request `vault.%s` for %s with key %s",
+            func,
+            minion_id,
+            private_key,
+        )
+        signature = base64.b64encode(salt.crypt.sign_message(private_key, minion_id))
+        result = __salt__["saltutil.runner"](
+            f"vault.{func}",
+            minion_id=minion_id,
+            signature=signature,
+            impersonated_by_master=True,
+            **kwargs,
+        )
+    return check_result(
+        result,
+        unwrap_client=unwrap_client,
+        unwrap_expected_creation_path=unwrap_expected_creation_path,
+    )
+
+
+def _get_event(opts):
+    if opts.get("__role") == "master":
+        return salt.utils.event.get_master_event(opts, opts["sock_dir"]).fire_event
+
+    global __salt__  # pylint: disable=global-statement
+    if not __salt__:
+        __salt__ = salt.loader.minion_mods(opts)
```

Comment:
@dwoz this looks a little funky to me. Do we want ot be adding `__salt__` to global here?

Reply:
Yeah, this is funky and definitely some kind of anti-pattern when it comes to the loader. Generally we can't rely on `__salt__` being available in utils. Although, I'm not sure this is fully documented anywhere so it's a reasonable mistake to make. What is the driver for trying to go through `__salt__` here instead of instantiating the event directly through `salt.utils.event.get_event`?

Reply:
Oh joy, I just went and looked at how many places we're using `__salt__` in utils and it's more than a few, about 50. Also, the current vault utils does the same thing as this.

Reply:
This is actually the result of me trying not to rely on `__salt__` being available since this module is always imported.
It is defined in the global scope directly: https://github.com/lkubb/salt/blob/32f977335765c530c66417a78504fdfde3dbdabd/salt/utils/vault/factory.py#L34-L35

I think I got this idea from the `botomod` and `namecheap` utils:
https://github.com/saltstack/salt/blob/eb71862449b93f496c678d892e0c0ad827278136/salt/utils/botomod.py#L67-L71
https://github.com/saltstack/salt/blob/eb71862449b93f496c678d892e0c0ad827278136/salt/utils/namecheap.py#L44-L47

Honestly, since writing this part of the code last year, I have learned a lot about Salt and the loader, so I might have written it this way because I didn't know better.
`__salt__` was initially used to get peer publishing going: https://github.com/lkubb/salt/blob/32f977335765c530c66417a78504fdfde3dbdabd/salt/utils/vault/factory.py#L754C25-L794 The event stuff I added later when `__salt__` was already part of the code.

I will check if the relevant modules can be imported directly, otherwise I might have to replicate the logic there. Would you mind giving me any pointers?

Reply:
I removed loading minion mods into the module's global namespace now. Since the `saltutil.runner`/`publish.runner` functions contain quite a bit of logic, I opted to import the modules directly and inject `__opts__` using `salt.utils.context.func_globals_inject`. The events are now created using `salt.utils.event.get_event`.

Is it OK to do it this way?